### PR TITLE
(JP-3015) Undersampling_correction step

### PR DESCRIPTION
**Pull request by dmggh**
Resolves [JP-3015](https://jira.stsci.edu/browse/JP-3015)

Closes #

This PR creates a new step in the detector1 pipeline between the jump and ramp_fitting steps to deal with the issues seen for undersampled NIRISS images. 

**Checklist for maintainers**
- [X] added entry in `CHANGES.rst` within the relevant release section
- [X] updated or added relevant tests
- [X] updated relevant documentation
- [X] added relevant milestone
- [X] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by dmggh** [comment]
All CI tests pass except the RTD build, which I could build successfully locally.

File: jwst/undersampling_correction/undersampling_correction.py
Comment (hbushouse) [inline jwst/undersampling_correction/undersampling_correction.py:39]: signal_threshold is a floating-point value, so does the use of "%s" formatting work properly for it?

**Review by hbushouse — COMMENTED** [review]
Given that the requirements for this also want us to change some behavior in the ramp_fit step, so that the readnoise variance is based on all groups for a pixel, including the ones that've been flagged by this step, we'll likely need to use a different DQ flag than DO_NOT_USE (we need a unique flag for this condition). I think there are 1 or 2 reserved/unused DQ values available yet, so we should think of an appropriate name and change this step code to use that new flag. Perhaps a name like "UNDERSAMP"?

File: docs/jwst/undersampling_correction/description.rst
Comment (dmggh) [reply docs/jwst/undersampling_correction/description.rst:4]: Fixed it.

File: docs/jwst/undersampling_correction/description.rst
Comment (dmggh) [reply docs/jwst/undersampling_correction/description.rst]: Done.

File: jwst/undersampling_correction/undersampling_correction.py
Comment (dmggh) [reply jwst/undersampling_correction/undersampling_correction.py]: Works as is, but changed for clarity.

File: jwst/undersampling_correction/undersampling_correction.py
Comment (hbushouse) [inline jwst/undersampling_correction/undersampling_correction.py:74]: While it's probably OK to just add the new dq value to the existing ones (because presumably that value won't exist yet for any pixels), I think it's safer to always use the `bitwise_or` function to manipulate flags, just in case. For example `gdq[wh_exc] = np.bitwise_or(gdq[wh_exec], dqflags.group['UNDERSAMP'])`

**Review by hbushouse — COMMENTED** [review]
Looks good. Just one requested change.

File: jwst/undersampling_correction/undersampling_correction.py
Comment (dmggh) [reply jwst/undersampling_correction/undersampling_correction.py]: Updated for consistency.

**Review by hbushouse — APPROVED** [review]

**Comment by hbushouse** [comment]
Has a regression test been run on this yet?

**Comment by dmggh** [comment]
No. I will run the reg tests against this branch.

**Comment by dmggh** [comment]
That comment should read "True", to match the correct code update.

File: jwst/lib/suffix.py
Comment (hbushouse) [inline jwst/lib/suffix.py:50]: I believe this entry needs to go into the calculated suffixes list starting down at line 60. This list (SUFFIXES_TO_ADD) is the list of standard suffixes for archived products, whereas the list below is for all the individual steps.

File: jwst/undersampling_correction/undersampling_correction.py
Comment (nden) [inline jwst/undersampling_correction/undersampling_correction.py:16]: Can you remove the `Short Summary` heading. Just keep the one line description below it.

File: jwst/undersampling_correction/undersampling_correction.py
Comment (nden) [inline jwst/undersampling_correction/undersampling_correction.py:52]: Same comment as above - remove the heading

**Comment by nden** [comment]
Question: The description says this step should be run on NIRISS data only but I did not see such condition implemented. Did I miss it or did the task change?

**Comment by hbushouse** [comment]
> Question: The description says this step should be run on NIRISS data only but I did not see such condition implemented. Did I miss it or did the task change?

Technically it's available for any instrument mode that wants to use it, but by default (at least initially) it should only be applied to NIRISS data, which will be controlled by parameter reference files. Note that the `skip` param in the step module has been hardwired to `True`, so it won't be executed for any exposure unless there's a parameter ref file override.

**Comment by nden** [comment]
As discussed offline, the current flagging of pixels above the threshold with `UNDERSAMP` will not achieve the desired slope computation, as these pixels will still be included in the computation. One way around this is to flag the pixels with `DO_NOT_USE` and `UNDERSAMP` so that they are excluded from the slope computation because of the `DO_NOT_USE` flag and the `UNDERSAMP` flag will be used in the variance computation. The code in ramp_fitting needs to be tracked to see if there will be any side affects.

File: docs/jwst/undersampling_correction/description.rst
Comment (nden) [inline docs/jwst/undersampling_correction/description.rst:31]: There's more to the algorithm and it should be fully described.

**Review by nden — CHANGES_REQUESTED** [review]
Decide if the current flagging is sufficient or it needs changes. Fully describe the algorithm in the RTD documentation.

**Comment by dmggh** [comment]
Reg test #507 ran without undersampling_correction-related errors : 
 https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/507/#showFailuresLink

It will be run again once additional changes to the flagging have been implemented.

File: jwst/undersampling_correction/undersampling_correction_step.py
Comment (hbushouse) [inline jwst/undersampling_correction/undersampling_correction_step.py:27]: Should add a log message here indicating that the step is being skipped because ngroups < 3.

File: jwst/undersampling_correction/undersampling_correction_step.py
Comment (nden) [inline jwst/undersampling_correction/undersampling_correction_step.py:30]: I'm not sure this should be a warning. Some users may interpret it as a bad thing. How about an `info` message. @bushouse?

Comment (nden) [reply jwst/undersampling_correction/undersampling_correction_step.py:30]: And can you add a test for this case - less than 3 groups.

**Review by nden — CHANGES_REQUESTED** [review]

**Comment by dmggh** [comment]
All CI tests pass except codecoverage and the RTD build, which I could build successfully locally.
The log for the RTD build seems to indicate a larger problem, and not related to my files :

WARNING: failed to reach any of the inventories with the following issues:
intersphinx inventory 'https://numpy.org/devdocs/objects.inv' not fetchable due to <class 'requests.exceptions.HTTPError'>: 404 Client Error: Not Found for url: https://numpy.org/devdocs/objects.inv

< snip >

intersphinx inventory has moved: http://matplotlib.org/objects.inv -> https://matplotlib.org/stable/objects.inv
[autosummary] generating autosummary for:
getting_started/contributing.rst,
getting_started/install.rst,
  etc.

File: CHANGES.rst
Comment (hbushouse) [inline CHANGES.rst:247]: Need to remove white space in the middle of `Detector1 Pipeline`

File: jwst/datamodels/schemas/core.schema.yaml
Comment (hbushouse) [inline jwst/datamodels/schemas/core.schema.yaml:2469]: I think I'd rather see the available characters get allocated towards more of the actual step name, rather than including "COR" (for correction) (most steps are corrections of some sort). How about `S_UNDSMP`?

File: jwst/pipeline/calwebb_detector1.py
Comment (hbushouse) [inline jwst/pipeline/calwebb_detector1.py:123]: The comment still implies NIRISS only. Is there some logic somewhere that applies that restriction? Is it in a pars ref file or something like that?

File: jwst/undersampling_correction/tests/test_undersampling_correction.py
Comment (hbushouse) [reply jwst/undersampling_correction/tests/test_undersampling_correction.py:131]: This still needs fixing. Should be "NIS" for the detector name.

File: jwst/undersampling_correction/undersampling_correction_step.py
Comment (hbushouse) [reply jwst/undersampling_correction/undersampling_correction_step.py:30]: I agree. Change it from warning to just `info`.

**Review by hbushouse — CHANGES_REQUESTED** [review]

File: CHANGES.rst
Comment (dmggh) [reply CHANGES.rst:247]: Done

File: jwst/datamodels/schemas/core.schema.yaml
Comment (dmggh) [reply jwst/datamodels/schemas/core.schema.yaml:2469]: Done

File: jwst/pipeline/calwebb_detector1.py
Comment (dmggh) [reply jwst/pipeline/calwebb_detector1.py:123]: No, there is no restriction. I removed it.

File: jwst/undersampling_correction/undersampling_correction_step.py
Comment (dmggh) [reply jwst/undersampling_correction/undersampling_correction_step.py:30]: Done

**Review by hbushouse — APPROVED** [review]
Code updates look good and CI results are fine (the 1 failure is just slowness of the system for a timed test). I hereby approve.

**Comment by dmggh** [comment]
Great, but merging is blocked until the failures are 'cleared'.

**Comment by hbushouse** [comment]
> Great, but merging is blocked until the failures are 'cleared'.

I'll merge once we're ready to reopen the jwst master branch after the B9.1 patch release.

File: jwst/undersampling_correction/tests/test_undersampling_correction.py
Comment (hbushouse) [inline jwst/undersampling_correction/tests/test_undersampling_correction.py:3]: I think these imports will still work, because they'll be automatically redirected to stdatamodels.jwst.datamodels, but we might as well go ahead and change them to point directly to the new location, as has been done in all the step codes.

**Review by hbushouse — COMMENTED** [review]
I still see code modules contained within jwst.datamodels in here. Shouldn't those have been removed via a rebase, because they're now actually contained in the stdatamodels PR?

File: jwst/undersampling_correction/tests/test_undersampling_correction.py
Comment (dmggh) [reply jwst/undersampling_correction/tests/test_undersampling_correction.py:3]: No errors were given when I did the rebase:
 (jwst_h) sal:jwst grumm$ git rebase  upstream/master 
Successfully rebased and updated refs/heads/master.

 And yes, I also thought that the jwst.datamodels would have been removed. I will look into this.

**Comment by hbushouse** [comment]
Hmmm, I wonder if the latest problems are due to the fact that this PR branch now completely deletes the jwst.datamodels.dqflags.py module, while master still has this file, but it's greatly modified compared to what it used to be before jwst.datamodels was moved over to stdatamodels. See https://github.com/spacetelescope/jwst/blob/master/jwst/datamodels/dqflags.py

So we still need to have that dqflags.py file in the code tree, it just needs to have the latest version of it from master.

**Comment by hbushouse** [comment]
OK, it looks like we're making progress here. RTD build is OK again, but now all of the jwst package tests fail due to the fact that the new DQ flag value is over in stdatamodels and we haven't done a new release of that yet. So this PR will have to wait for final testing until we get a new release of stdatamodels.

**Comment by hbushouse** [comment]
@dmggh stdatamodels 1.2.0 has been released, which contains the DQ-related changes to support this PR. So please go ahead and add a modification to `setup.cfg` for the stdatamodels upgrade (i.e. at https://github.com/spacetelescope/jwst/blob/master/setup.cfg#L47) to force it to use 1.2.0 or later, and add that change to this PR. That should allow the CI tests to succeed (I hope).

**Comment by dmggh** [comment]
Closing this PR (7396) as it has been superseded by #7479.

**Pull request closed without merging**
